packages/contracts/contracts/ChugSplashRegistry.sol

```diff
@@ -1,50 +1,312 @@
+     * Registers a new project.
      *
      * @param _name Name of the new ChugSplash project.
-     * @param _owner Initial owner for the new ChugSplashManager contract.
-     * @return Address of the newly deployed ChugSplashManager contract.
+     * @param _manager Initial manager for the new project.
+     * @return Address of the manager for the new project.
      */
-    function register(string memory _name, address _owner) public returns (address) {
+    function register(string memory _name, address _manager) public returns (address) {
         // TODO: Standardize error reporting system.
         require(
-            address(registry[_name]) == address(0),
+            projects[_name].manager == address(0),
             "ChugSplashRegistry: name already registered"
         );
 
-        // By making the salt be the hash of the name, we can easily predict the address of the
-        // ChugSplashManager contract based on the project's name.
-        bytes32 salt = keccak256(bytes(_name));
-        registry[_name] = new ChugSplashManager{ salt: salt }(_owner);
-        emit ChugSplashProjectRegistered(_name, msg.sender, _owner, address(registry[_name]));
-        return address(registry[_name]);
+        ChugSplashProject storage project = projects[_name];
+        project.manager = _manager;
+        emit ChugSplashProjectRegistered(_name, msg.sender, _manager);
+
+        // TODO: We used to return the ChugSplashManager address here. Should we keep this new
+        // return value or remove it?
+        return _manager;
+    }
+
+    /**
+     * Allows a manager to propose a new ChugSplash bundle to be executed.
+     *
+     * @param _name Name of the ChugSplash project.
+     * @param _bundleHash Hash of the bundle to execute.
+     * @param _bundleSize Total number of actions in the bundle.
+     * @param _configUri URI pointing to the config file for the bundle.
+     */
+    function proposeChugSplashBundle(
+        string memory _name,
+        bytes32 _bundleHash,
+        uint256 _bundleSize,
+        string memory _configUri
+    ) public onlyManager(_name) {
+        ChugSplashProject storage project = projects[_name];
+        require(
+            project.status == ChugSplashBundleStatus.EMPTY,
+            "ChugSplashRegistry: bundle already exists"
+        );
+
+        project.status = ChugSplashBundleStatus.PROPOSED;
+        project.executions = new bool[](_bundleSize);
+
+        emit ChugSplashBundleProposed(_name, _bundleHash, _bundleSize, _configUri);
+    }
+
+    /**
+     * Allows a manager to approve a bundle to be executed. Note that the bundle can be executed
+     * as soon as the bundle is approved.
+     *
+     * @param _name Name of the ChugSplash project.
+     * @param _bundleHash Hash of the bundle to approve.
+     */
+    function approveChugSplashBundle(string memory _name, bytes32 _bundleHash) public onlyManager(_name) {
+        ChugSplashProject storage project = projects[_name];
```

Comment:
Assuming we have multiple proposed bundles, we'll also want to check here that the `_bundleHash` corresponds to some bundle that has been previously proposed.